src/wallets/bitcoin.js

```diff
@@ -0,0 +1,510 @@
+        network: NETWORK
+      }),
+      network: NETWORK
+    })
+    return address
+  }
+
+  xPrivSigner = (
+    inputs: Array<Object>,
+    to: string,
+    // value to be sent
+    satoshiValue: number,
+    // total fee in satoshi
+    fee: number,
+    changeIndex: number
+  ) => {
+    // use the first account
+    let account = this.getAccount()
+
+    const keyPair = bitcoin.ECPair.fromPrivateKey(account.privateKey)
+    const p2wpkh = bitcoin.payments.p2wpkh({
+      pubkey: keyPair.publicKey,
+      network: NETWORK
+    })
+
+    const { address } = bitcoin.payments.p2sh({
+      redeem: p2wpkh,
+      network: NETWORK
+    })
+
+    const p2sh = bitcoin.payments.p2sh({ redeem: p2wpkh, network: NETWORK })
+    const txb = new bitcoin.TransactionBuilder(NETWORK)
+
+    // add all inputs
+    inputs.map(input => txb.addInput(input.txHash, input.outputIndex))
+
+    // the actual "spend"
+    txb.addOutput(to)
+
+    // change
+    const inputValueTotal = inputs.reduce((total, input) => total + input.value)
+    const change = inputValueTotal - satoshiValue - fee
+    txb.addOutput(address, change)
+
+    // sign with first address's privateKey
+    txb.sign(0, keyPair, p2sh.redeem.output, null, inputValueTotal)
+
+    const tx = txb.build()
+
+    // return raw tx in hex
+    return tx.toHex()
+  }
+
+  broadcastBtcRawTx = async (txRaw: string) => {
+    const rv = await axios.post(`${url.LEDGER_API_URL}/transactions/send`, {
+      tx: txRaw
+    })
+    return rv.data.result
+  }
+
+  collectUtxos = (
+    addressPool: Array<AddressBitcoin>,
+    value: BasicTokenUnit = '0',
+    txFeePerByte: number = 15
+  ) => {
+    let utxosCollected = []
+    let valueCollected = new BN(0)
+    let i = 0
+    let size = 0
+    let fee = new BN(0)
+    while (i < addressPool.length) {
+      let utxos = addressPool[i].utxos
+      for (let j = 0; j < utxos.length; j++) {
+        const utxo = utxos[j]
+        utxosCollected.push({
+          ...utxo,
+          keyPath: addressPool[i].path
+        })
+        size = this._estimateTransactionSize(utxosCollected.length, 2, true).max
+        fee = new BN(size).mul(new BN(txFeePerByte))
+        valueCollected = valueCollected.add(new BN(utxo.value))
+        if (valueCollected.gte(new BN(value).add(fee))) {
+          return {
+            fee: fee.toString(),
+            size,
+            utxosCollected
+          }
+        }
+      }
+      i += 1
+    }
+    console.warn('Transfer amount greater and fee than utxo values.')
+    return {
+      fee: fee.toString(),
+      size,
+      utxosCollected
+    }
+  }
+
+  _discoverAddress = async (
+    xpub: string,
+    accountIndex: number,
+    change: number,
+    offset: number,
+    progress: ?Function
+  ): Promise<{ nextIndex: number, addresses: Array<Address> }> => {
+    let gap = 0
+    let addresses: Array<AddressBitcoin> = []
+    let currentIdx = offset
+    let lastUsedIdx = offset - 1
+    while (gap < 5) {
```

Comment:
should we consider changing the gap back to 20?

Reply:
fixed